federation-js/src/composition/utils.ts

```diff
@@ -62,20 +62,39 @@ export function mapFieldNamesToServiceName<Node extends { name: NameNode }>(
 
 export function findDirectivesOnNode(
   node: Maybe<
-    | TypeDefinitionNode
-    | TypeExtensionNode
     | FieldDefinitionNode
-    | SchemaDefinitionNode
+    | FieldNode
+    | DefinitionNode
   >,
   directiveName: string,
 ) {
-  return node && node.directives
+  return node && 'directives' in node && node.directives
     ? node.directives.filter(
         directive => directive.name.value === directiveName,
       )
     : [];
 }
 
+export function findSelectionSetOnNode(
+  node: Maybe<
+    | FieldDefinitionNode
+    | DefinitionNode
+  >,
+  directiveName: string,
+  elementInSelectionSet: string,
+) {
+  // assume there is only one directive with this name that selects this element
+  return node && 'directives' in node && node.directives
+    ? node.directives.find(
+        directive =>
+          directive.name.value === directiveName && directive.arguments?.some(
+            argument => argument.value.kind === "StringValue" &&
+              argument.value.value.includes(elementInSelectionSet)
+          ))?.arguments?.find(
+            argument => argument.name.value === 'fields')?.value
+    : undefined;
```

Comment:
1) we should use `Kind.STRING` for `"StringValue"` (or use the `isStringValueNode` predicate fn!)
2) I'm not totally sure I understand this correctly but I think I do. Is the `includes` here possibly error-prone?

I'm imagining 2 keys:
`@key(fields: "{ id }")`
`@key(fields: "{ sku { id } }")`

Could this scenario break this implementation? I might be mistaken and just need to understand this a bit better.

Reply:
In the scenario you described:

`type SomeType @key(fields: "id") @key(fields: "sku { id }") { ` and we are looking for the field with `id` on it, yeah this would break. If this were the case, and we wanted to find a top level field, would it be sufficient to make sure that string is not wrapped in curly braces? I'll write a test for this

Reply:
The resolution here was to search for the entire selection set, so we know we are selecting the correct key.